### PR TITLE
Support for DigiByte(DGB)

### PR DESCRIPTION
**Pull request by johnnynanjiang**
## Questions
1. Could CoinType+Address.swift be auto generated from TWHRP.cpp?

2. Set the followings values as per advice from DigiByte community, address verified against iancoleman.io

```
        "derivationPath": "m/84'/20'/0'/0/0",
        "xpub": "xpub",
        "xprv": "xprv",
```

3. Any advice for testing for TX signing and broadcasting?

4. A concern from DigiByte community, any advice?

```
Quick run-down, no doubt you're aware that we're bringing out DigiAssets shortly.
Right now, most 3rd-party wallets use "Bitcoin" as the seed.
The DigiByte mobile apps use "DigiByte", naturally making them incompatible when restoring, say, with Coinomi for example.
What we're looking to do is get wallet vendors who specifically support DigiAssets, or at least are aware of DigiAssets, to use the "DigiByte" seed, rather than "Bitcoin". The reason for this is if we restore a seed phrase to a DigiByte wallet that does *not* support DigiAssets, when you gather the inputs if you gather the inputs that have DigiAssets associated with them, and spend those, you effectively burn the asset.
```

5. Ran gradle build in Android and got the following errors, any advice?

```
Errors found:
  
  /Users/nanjiang/.gradle/caches/modules-2/files-2.1/io.grpc/grpc-core/1.9.0/cf76ab13d35e8bd5d0ffad6d82bb1ef1770f050c/grpc-core-1.9.0.jar: Error: Invalid package reference in library; not included in Android: javax.naming.directory. Referenced from io.grpc.internal.DnsNameResolver.JndiResolver. [InvalidPackage]
  /Users/nanjiang/.gradle/caches/modules-2/files-2.1/io.grpc/grpc-core/1.9.0/cf76ab13d35e8bd5d0ffad6d82bb1ef1770f050c/grpc-core-1.9.0.jar: Error: Invalid package reference in library; not included in Android: javax.naming. Referenced from io.grpc.internal.DnsNameResolver.JndiResolver. [InvalidPackage]
```

### CODE REVIEW

File: coins.json [hunk 1]
```diff
@@ -87,13 +87,13 @@
...
         "blockchain": "Bitcoin",
-        "derivationPath": "m/44'/20'/0'/0'",
+        "derivationPath": "m/84'/20'/0'/0/0",
         "curve": "secp256k1",
         "publicKeyType": "secp256k1",
         "publicKeyHasher": "sha256ripemd",
         "base58Hasher": "sha256d",
-        "xpub": "zpub",
-        "xprv": "zprv",
+        "xpub": "xpub",
```
Comment (hewigovens) [inline coins.json]: bip44 -> xpub
bip84 -> zpub

File: include/TrustWalletCore/TWP2PKHPrefix.h [hunk 2]
```diff
@@ -17,7 +17,7 @@ enum TWP2PKHPrefix {
     TWP2PKHPrefixBitcoin = 0x00,
     TWP2PKHPrefixDash = 0x4C,
     TWP2PKHPrefixDecred = 0x3f,
-    TWP2PKHPrefixDogecoin = 0x1e,
+    TWP2PKHPrefixDogecoin_DigiByte = 0x1e,
```
Comment (hewigovens) [inline include/TrustWalletCore/TWP2PKHPrefix.h]: or maybe just `TWP2PKHPrefixD`? and add comment // for Doge/DigiByte etc

File: coins.json (hunk 1 above)
Comment (johnnynanjiang) [reply coins.json]: Set the followings values as per advice from DigiByte community, address verified against iancoleman.io

```
        "derivationPath": "m/84'/20'/0'/0/0",
        "xpub": "xpub",
        "xprv": "xprv",
```

Comment (hewigovens) [reply coins.json]: But it doesn’t make sense, if you use native segwit address, you should use zpub

Comment (johnnynanjiang) [reply coins.json]: I think you are right that `BIP84` uses `zpub` \ `zprv`.
For some reason I got told to use `xpub` \ `xprv` by the DigiByte community, and https://iancoleman.io/bip39/ also uses `xpub` \ `xprv` for DGB segwit addresses, that is where I got source of truth for our tests.

File: include/TrustWalletCore/TWP2PKHPrefix.h (hunk 2 above)
Comment (johnnynanjiang) [reply include/TrustWalletCore/TWP2PKHPrefix.h]: Done in 84bc3aa

File: swift/Tests/HDWalletTests.swift [hunk 3]
```diff
@@ -53,6 +53,15 @@ class HDWalletTests: XCTestCase {
         XCTAssertEqual("bc1qumwjg8danv2vm29lp5swdux4r60ezptzz7ce85", address)
     }
 
+    func testDeriveDigiByte() {
+        let bitcoin = CoinType.digiByte
+        let wallet = HDWallet.test
+        let key = wallet.getKeyForCoin(coin: bitcoin)
```
Comment (vikmeup) [inline swift/Tests/HDWalletTests.swift]: `(coin: .digiByte)`

Comment (johnnynanjiang) [reply swift/Tests/HDWalletTests.swift]: Done in 58f703b0ae909c814e1803b3e7e99ccc83e61fbc

File: coins.json (hunk 1 above)
Comment (johnnynanjiang) [reply coins.json]: Done in 1f5f21b6d8896693b9e8bbcb4354012f19e9799e

**Comment by johnnynanjiang** [comment]
PR updated, please have a review, I will continue on tx verification as discussed, @vikmeup @hewigovens

**Comment by hewigovens** [comment]
@johnnynanjiang here is a bitcoin fork check list, can you make sure these are all covered?

- [ ] derive bip44 p2pkh address or bip84 bech32 address
- [ ] validate bip44 / bip49 / bip84 addresses (p2pkh / p2sh prefix, hrp)
- [ ] check xpub / xprv version
- [ ] cross check xpub / xprv with ledger / trezor
- [ ] derive address from xpub (random index)
- [ ] check lock script for  bip44 / bip49 / bip84 addresses (can do it on block explorer)
- [ ] check tx detail url
- [ ] check fee preference, estimatefee/2 or static
- [ ] add signer test

**Comment by johnnynanjiang** [comment]
The following has been done as part of this first PR for DigiByte support

[ ] derive bip44 p2pkh address or bip84 bech32 address
[ ] validate bip44 / bip49 / bip84 addresses (p2pkh / p2sh prefix, hrp)
[ ] check xpub / xprv version
[ ] cross check xpub / xprv with ledger / trezor
[ ] derive address from xpub (random index)

File: src/Coin.cpp [hunk 4]
```diff
@@ -195,14 +196,20 @@ std::string TW::deriveAddress(TWCoinType coin, const PublicKey& publicKey) {
...
 
     case TWCoinTypeDecred:
         return Decred::Address(publicKey).string();
 
     case TWCoinTypeDogecoin:
-        return Bitcoin::Address(publicKey, TWP2PKHPrefixDogecoin).string();
+        return Bitcoin::Address(publicKey, TWP2PKHPrefixD).string();
+
+    case TWCoinTypeDigiByte:
+        return Bitcoin::SegwitAddress(publicKey, 0, HRP_DIGIBYTE).string();
```
Comment (vikmeup) [inline src/Coin.cpp]: I think most exchanges do not support bech32 address for DG, should we use `Bitcoin::Address(publicKey, TWP2PKHPrefixD).string();` instead? 

@hewigovens what do you think?

Comment (hewigovens) [reply src/Coin.cpp]: I suggest to ask DigiByte team. GRS/VIA want bip84 and Qtum wants bip44

Comment (johnnynanjiang) [reply src/Coin.cpp]: Fair point, actually DigiByte guys suggested BIP84 when I talked to them few days ago

**Comment by hewigovens** [comment]
address starts with `S` is also valid, can you add test case to cover that? address validation and lock script

https://dgb2.trezor.io/tx/965eb4afcd0aa6e3f4f8fc3513ca042f09e6e2235367fa006cbd1f546c293a2a

**Comment by johnnynanjiang** [comment]
> address starts with `S` is also valid, can you add test case to cover that? address validation and lock script
> 
> https://dgb2.trezor.io/tx/965eb4afcd0aa6e3f4f8fc3513ca042f09e6e2235367fa006cbd1f546c293a2a
> 
> > **[TREZOR DigiByte Explorer](https://dgb2.trezor.io/tx/965eb4afcd0aa6e3f4f8fc3513ca042f09e6e2235367fa006cbd1f546c293a2a)**TREZOR DigiByte Explorer

Done in fba9c10a0a7229068610a0190185728833e9a3d8, also added a test for P2WPKH.

File: src/Coin.cpp [hunk 5]
```diff
@@ -74,7 +74,11 @@ bool TW::validateAddress(TWCoinType coin, const std::string& string) {
         return Decred::Address::isValid(string);
 
     case TWCoinTypeDogecoin:
-        return Bitcoin::Address::isValid(string, {{TWP2PKHPrefixDogecoin}, {TWP2SHPrefixDogecoin}});
+        return Bitcoin::Address::isValid(string, {{TWP2PKHPrefixD}, {TWP2SHPrefixDogecoin}});
+
+    case TWCoinTypeDigiByte:
+        return Bitcoin::SegwitAddress::isValid(string, HRP_DIGIBYTE) ||
+               Bitcoin::Address::isValid(string, {{TWP2PKHPrefixD}, {TWP2SHPrefixBitcoin}});
```
Comment (hewigovens) [inline src/Coin.cpp]: We need also to add p2sh prefix S here

Comment (johnnynanjiang) [reply src/Coin.cpp]: Done in 2f76ee734e60a9d39c75665fe54524b90e1b924c

**Pull request merged**